pkg/sql/catalog/hydratedtables/hydratedcache.go

```diff
@@ -0,0 +1,265 @@
+// been used in a long time.
+
+// Cache caches table descriptors which have their user-defined types hydrated.
+// The cache's contract is a bit tricky. In order to use a hydrated type, the
+// caller needs to have a lease on the relevant type descriptor. The way that
+// this is made to work is that the user provides a handle to a leased
+// ImmutableCopy and then the cache will call a function for each
+// of the
+type Cache struct {
+	settings *cluster.Settings
+	g        singleflight.Group
+	metrics  Metrics
+	mu       struct {
+		syncutil.RWMutex
+		cache *cache.UnorderedCache
+	}
+}
+
+// Metrics returns the cache's metrics.
+func (c *Cache) Metrics() *Metrics {
+	return &c.metrics
+}
+
+var _ metric.Struct = (*Metrics)(nil)
+
+// Metrics exposes cache metrics.
+type Metrics struct {
+	Hits   *metric.Counter
+	Misses *metric.Counter
+}
+
+func makeMetrics() Metrics {
+	return Metrics{
+		Hits:   metric.NewCounter(metaHits),
+		Misses: metric.NewCounter(metaMisses),
+	}
+}
+
+// MetricStruct makes Metrics a metric.Struct.
+func (m *Metrics) MetricStruct() {}
+
+var (
+	metaHits = metric.Metadata{
+		Name:        "sql.hydrated_table_cache.hits",
+		Help:        "counter on the number of cache hits",
+		Measurement: "reads",
+		Unit:        metric.Unit_COUNT,
+		MetricType:  io_prometheus_client.MetricType_COUNTER,
+	}
+	metaMisses = metric.Metadata{
+		Name:        "sql.hydrated_table_cache.misses",
+		Help:        "counter on the number of cache misses",
+		Measurement: "reads",
+		Unit:        metric.Unit_COUNT,
+		MetricType:  io_prometheus_client.MetricType_COUNTER,
+	}
+)
+
+// CacheSize controls the size of the LRU cache.
+var CacheSize = settings.RegisterNonNegativeIntSetting(
+	"sql.catalog.hydrated_tables.cache_size",
+	"number of table descriptor versions retained in the hydratedtables LRU cache",
+	128)
+
+// NewCache constructs a new Cache.
+func NewCache(settings *cluster.Settings) *Cache {
+	c := &Cache{
+		settings: settings,
+		metrics:  makeMetrics(),
+	}
+	c.mu.cache = cache.NewUnorderedCache(cache.Config{
+		Policy: cache.CacheLRU,
+		ShouldEvict: func(size int, key, value interface{}) bool {
+			return size > int(CacheSize.Get(&settings.SV))
+		},
+	})
+	return c
+}
+
+type hydratedTableDescriptor struct {
+	tableDesc *tabledesc.Immutable
+	typeDescs []*cachedType
+}
+
+type cachedType struct {
+	name    tree.TypeName
+	typDesc catalog.TypeDescriptor
+}
+
+// GetHydratedTableDescriptor returns an ImmutableCopy with the types
+// hydrated. It may use a cached copy but all of the relevant type descriptors
+// will be retrieved via the TypeDescriptorResolver.
+func (c *Cache) GetHydratedTableDescriptor(
+	ctx context.Context, table *tabledesc.Immutable, lookupFunc catalog.TypeDescriptorResolver,
+) (_ *tabledesc.Immutable, err error) {
+	if !table.ContainsUserDefinedTypes() {
+		return table, nil
+	}
+
+	// TODO(ajwerner): This cache may thrash a bit right when a version of a type
```

Comment:
That's what I mean by finer granularity. It will avoid the thrashing but it's a good deal more complex. In particular, it's more complex w.r.t. the cache. I'll do it as a follow-up if you'll allow it.